### PR TITLE
fix: Sanitise marketplace name to kebab-case in output mappers

### PR DESCRIPTION
**Pull request by sergio-sisternes-epam**
## Description

Repository names containing dots, underscores, or other non-kebab-case characters (e.g. `my.marketplace`) flow into `marketplace.json` verbatim. The Copilot App rejects these with a validation error: "name Marketplace name must be kebab-case (letters, numbers, and hyphens)".

Since we cannot control how users name their repositories, this PR adds a `sanitise_marketplace_name()` function that converts names to kebab-case at the output boundary -- when composing `marketplace.json`. The function lowercases, replaces non-alphanumeric characters with hyphens, collapses consecutive hyphens, and strips leading/trailing hyphens.

Applied in both `ClaudeMarketplaceMapper` and `CodexMarketplaceMapper`. Internal resolution, registry lookups, and install flows remain untouched -- the original name is preserved everywhere except the consumer-facing JSON `"name"` field. The Codex `displayName` also keeps the original for human readability.

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Documentation
- [ ] Maintenance / refactor

## Testing

- [x] Tested locally
- [x] All existing tests pass (1409 marketplace unit tests)
- [x] Added tests for new functionality (`tests/unit/marketplace/test_name_sanitisation.py` -- 16 parametrised cases covering dots, underscores, spaces, slashes, special characters, mixed case, edge cases, and fallback behaviour)

## Spec conformance (OpenAPM v0.1)

- [x] N/A -- this PR does not change OpenAPM-observable behaviour.

### CODE REVIEW

File: src/apm_cli/marketplace/output_mappers.py [hunk 1]
```diff
@@ -15,6 +16,25 @@
 from .diagnostics import BuildDiagnostic
 from .errors import BuildError
 
+
+def sanitise_marketplace_name(name: str) -> str:
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/marketplace/output_mappers.py]: Good catch -- renamed to `sanitize_marketplace_name` and `sanitized` for the local variable to match the codebase convention (`_sanitize_bundle_name`, `_sanitize_cache_name`). Fixed in 0e40142e.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/apm_cli/marketplace/output_mappers.py [hunk 2]
```diff
@@ -15,6 +16,25 @@
...
+    """Convert a marketplace name to kebab-case for Copilot App compatibility.
+
+    Repo names like ``my.marketplace`` or ``My_Package`` are valid on GitHub
+    but rejected by the Copilot App which requires kebab-case
+    (lowercase letters, digits, and hyphens only).
+
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/marketplace/output_mappers.py]: Updated the docstring to clarify this is output-boundary normalization of the marketplace identity (from `config.name`), not limited to repo names. Fixed in 0e40142e.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

**Comment by danielmeppiel** [comment]
## APM Review Panel: `ship_now`

> Ship: kebab-case sanitiser closes the silent-rejection cliff for marketplace producers with proven mutation-break coverage and zero blocking findings.

cc @sergio-sisternes-epam @danielmeppiel -- a fresh advisory pass is ready for your review.

All nine panelists converged on ship. The two active architectural voices (python-architect, cli-logging-expert) confirmed the shared-helper extraction is well-motivated and follows the existing collect-then-render diagnostic pattern. Supply-chain-security certified a net-positive: the sanitiser narrows the injection surface on emitted names while the raw name stays internal-only. The test-coverage-expert returned zero findings with six evidence blocks at integration-with-fixtures tier, all outcome=passed, and the mutation-break gate is proven -- removing the sanitize guard fails name-assertion tests; neutering the diagnostic condition fails `test_rewrite_emits_warning_diagnostic`. This is the empirical-proof bar the maintainer asked for.

No material dissent exists. The two deferred nits are cleanly out of scope: (1) typing `BuildDiagnostic.level` as `Literal` touches a pre-existing shared type across the codebase, not this PR's surface; (2) widening the compose-only preview helper to surface diagnostics is a separate refactor since the real write path already renders them. Both are bounded follow-ups with no user-facing regression risk if deferred.

Strategically, this change reinforces APM's "just works" positioning for the producer publish flow. A marketplace.json that silently fails on the Copilot App side is exactly the kind of first-run cliff that kills adoption before a user ever sees value. Closing it with a warn-not-error approach (npm-like ergonomics per devx-ux-expert) is the right trade-off: producers are informed, never blocked.

**Aligned with:** Pragmatic as npm (warn-not-error mirrors npm normalize-package-name; producers informed, never blocked); Portable by manifest (raw manifest name preserved for internal resolution, only the emitted artifact is sanitised); Multi-harness / multi-host (Claude and Codex mappers apply identical sanitisation via one shared helper); OSS community-driven (real community pain point shipped with full empirical proof); Secure by default (emitted name constrained to `[a-z0-9-]`, closing injection surface on the published artifact).

**Growth signal.** Closes the "marketplace.json silently rejected by Copilot App" first-run cliff for producers. Amplify in release notes as a "just works" DX win -- producers no longer need to manually kebab-case their config name to get a successful publish.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 0 | 2 | Architecturally clean output-boundary sanitization; shared-helper extraction well-motivated; diagnostic aggregation follows existing collect-then-render pattern. |
| CLI Logging Expert | 0 | 0 | 1 | Diagnostic wording clear, actionable, ASCII `[!]`-prefixed, names both values; `level=warning` correct for a silent rewrite. |
| DevX UX Expert | 0 | 0 | 1 | Warn-not-error is defensible npm-like ergonomics; fallback and diagnostic UX solid. |
| Supply Chain Security Expert | 0 | 0 | 1 | Net-positive: sanitiser closes injection surface on emitted name; no blocking supply-chain risk. |
| OSS Growth Hacker | 0 | 0 | 2 | Removes a silent producer-funnel footgun; docs + diagnostic coverage solid. |
| Doc Writer | 0 | 0 | 2 | Prose matches sanitiser; anchor `#72-block-fields` resolves; two discoverability nits. |
| Test Coverage Expert | 0 | 0 | 0 | All behavioral surfaces have regression-trap tests at integration-with-fixtures tier; 22 tests pass; mutation-break confirmed. Ship. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships. All eight in-scope nits were folded into this PR (see "Folded in this run"); two out-of-scope nits are deferred.

### Top 2 follow-ups

1. **[Python Architect]** Type `BuildDiagnostic.level` as a `Literal` -- pre-existing shared type used across the codebase; typing it narrows the contract but touches many call-sites. Bounded refactor, no user-facing urgency.
2. **[CLI Logging Expert]** Surface diagnostics from the compose-only `compose_codex_marketplace_json` helper -- it returns a 2-tuple and drops the diagnostic list. The real write path already surfaces them; widening the preview helper is a separate refactor with no regression risk if deferred.

### Architecture

```mermaid
classDiagram
    direction LR
    class MarketplaceOutputMapper {
      <<AbstractBase>>
      +compose(config, resolved, remote_metadata) MapperResult
    }
    class ClaudeMarketplaceMapper {
      <<ConcreteStrategy>>
      +compose(...) MapperResult
    }
    class CodexMarketplaceMapper {
      <<ConcreteStrategy>>
      +compose(...) MapperResult
    }
    class MapperResult {
      <<ValueObject>>
      +document dict
      +warnings tuple
      +diagnostics tuple~BuildDiagnostic~
    }
    class BuildDiagnostic {
      <<ValueObject>>
      +level str
      +message str
    }
    class MarketplaceBuilder {
      <<Facade>>
      +build() BuildReport
    }
    class BuildReport {
      <<ValueObject>>
      +outputs tuple~MarketplaceOutputReport~
      +diagnostics tuple~BuildDiagnostic~
    }
    class sanitize_marketplace_name {
      <<Pure>>
      +__call__(name) str
    }
    class _sanitized_name_with_diagnostic {
      <<Pure>>
      +__call__(config_name) tuple
    }
    MarketplaceOutputMapper <|-- ClaudeMarketplaceMapper
    MarketplaceOutputMapper <|-- CodexMarketplaceMapper
    ClaudeMarketplaceMapper ..> MapperResult : returns
    CodexMarketplaceMapper ..> MapperResult : returns
    MapperResult *-- BuildDiagnostic : contains
    MarketplaceBuilder *-- MarketplaceOutputMapper : delegates
    MarketplaceBuilder ..> BuildReport : returns
    BuildReport o-- BuildDiagnostic : aggregates
    note for _sanitized_name_with_diagnostic "Shared helper: calls sanitize_marketplace_name,\nemits BuildDiagnostic when rewrite occurs"
    note for MarketplaceBuilder "Collect-then-render: diagnostics\naccumulate across mappers into BuildReport"
    class ClaudeMarketplaceMapper:::touched
    class CodexMarketplaceMapper:::touched
    class sanitize_marketplace_name:::touched
    class _sanitized_name_with_diagnostic:::touched
    classDef touched fill:#fff3b0,stroke:#d47600
```

```mermaid
flowchart TD
    A["apm.yml: name = 'My.Marketplace_Name'"] --> B["MarketplaceBuilder.build()"]
    B --> C["ClaudeMarketplaceMapper.compose() / CodexMarketplaceMapper.compose()"]
    C --> D["_sanitized_name_with_diagnostic(config.name)"]
    D --> E["sanitize_marketplace_name(name)"]
    E --> F{"sanitized == config_name?"}
    F -- Yes --> G["Return sanitized, diagnostics=[]"]
    F -- No --> H["Emit BuildDiagnostic level=warning\noriginal + sanitized in message"]
    H --> I["Return sanitized, diagnostics=[diag]"]
    G --> J["doc['name'] = sanitized"]
    I --> J
    J --> K{"Mapper type?"}
    K -- Claude --> L["diagnostics = list(name_diagnostics) +\nper-plugin diagnostics"]
    K -- Codex --> M["doc['interface']['displayName'] = config.name\nMapperResult(doc, (), tuple(name_diagnostics))"]
    L --> N["MapperResult(doc, warnings, tuple(diagnostics))"]
    M --> O["[FS] MarketplaceBuilder writes marketplace.json"]
    N --> O
    O --> P["BuildReport.diagnostics aggregates all outputs"]
```

### Recommendation

Zero blocking findings, zero recommended findings, all in-scope nits folded, mutation-break gate proven, lint silent, 1484+22 tests green, and test-coverage-expert certified ship at integration-with-fixtures tier. The two deferred nits are cleanly out of scope with no user-facing risk. Ship.

---

### Folded in this run

- (user-request) e2e integration test proving the full build emits a sanitized kebab-case name (empirical proof) -- resolved in `151fdd5d`.
- (shepherd) CHANGELOG `[Unreleased] > Fixed` entry for the behavior change -- resolved in `b7139be8`.
- (devx-ux-expert + cli-logging-expert) non-blocking build warning naming original + emitted name on rewrite, via a shared `_sanitized_name_with_diagnostic` helper wired into both mappers -- resolved in `52ca9b51`.
- (doc-writer) docs drift: publish guide + manifest schema now describe the kebab-case normalisation -- resolved in `52ca9b51`.
- (python-architect) pre-compile the sanitiser regexes at module scope (codebase idiom) -- resolved in `3a3ba870`.
- (devx-ux-expert) warning cited `marketplace.yml`; retargeted to "your marketplace config" since the recommended location is the `apm.yml` `marketplace:` block -- resolved in `3a3ba870`.
- (supply-chain-security-expert) documented the intentional many-to-one mapping in the sanitiser docstring -- resolved in `3a3ba870`.
- (oss-growth-hacker) CHANGELOG re-led with user value; publish guide now names the rewrite warning -- resolved in `3a3ba870`.
- (doc-writer) added a kebab-case pointer on the schema 3.1 `name` row; split the dense producer sentence -- resolved in `3a3ba870`.

### Copilot signals reviewed

- `src/apm_cli/marketplace/output_mappers.py` -- LEGIT (sanitise -> sanitize naming): already resolved by the author at HEAD `0e40142e` before this shepherding pass.
- `src/apm_cli/marketplace/output_mappers.py` -- LEGIT (docstring wording): already resolved by the author at HEAD `0e40142e` before this shepherding pass.

### Deferred (out-of-scope follow-ups)

These items were surfaced by the panel but cross the stated scope of this PR. Each suggests a separate follow-up.

- (python-architect) Type `BuildDiagnostic.level` as `Literal["warning", "verbose"]` -- scope boundary: a pre-existing shared type in `diagnostics.py` used across the marketplace subsystem, outside this PR's name-sanitisation surface.
- (cli-logging-expert) Widen the compose-only `compose_codex_marketplace_json` helper to surface diagnostics -- scope boundary: the user-facing write path (`write_codex_marketplace_json` -> `write_output` -> `compose_output`) already surfaces the diagnostic; changing the preview-helper signature is a separate refactor.

### Regression-trap evidence (mutation-break gate)

- `tests/integration/marketplace/test_name_sanitisation_e2e.py::TestClaudeOutputSanitisesName::test_written_json_name_is_kebab_case` -- removed the `sanitize_marketplace_name` guard from both mappers; the name-assertion tests FAILED as expected; guard restored.
- `tests/integration/marketplace/test_name_sanitisation_e2e.py::TestClaudeOutputSanitisesName::test_rewrite_emits_warning_diagnostic` -- neutered the `if sanitized != config_name` diagnostic condition; the diagnostic test FAILED as expected; condition restored.

### Lint contract

`uv run --extra dev ruff check src/ tests/` and `uv run --extra dev ruff format --check src/ tests/` both silent. Pylint R0801 duplication guard: 10.00/10.

### CI

All required checks green on `3a3ba870`: Lint, Build & Test Shard 1 (Linux), Build & Test Shard 2 (Linux), gate, Coverage Combine (Linux), PR Binary Smoke (Linux), Spec conformance gate, CodeQL (after 0 CI fix iterations).

### Mergeability status

| PR | head SHA | CEO stance | iters | folds | defers | Copilot rounds | CI | mergeable | mergeStateStatus | notes |
|----|----------|------------|-------|-------|--------|----------------|----|-----------|------------------|-------|
| #2008 | `3a3ba870` | ship_now | 1 | 9 | 2 | 1 | green | MERGEABLE | BLOCKED | BLOCKED = awaiting required review approval; CI is green, no merge conflict. |

### Convergence

1 outer iteration; 1 Copilot round. Final panel stance: `ship_now`.

Ready for maintainer review.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect

- **[nit]** Pre-compile the two regexes used in `sanitize_marketplace_name` at `src/apm_cli/marketplace/output_mappers.py:38`
  The two `re.sub` calls re-compile on every invocation; the idiomatic APM pattern is module-level pre-compiled patterns. FOLDED in `3a3ba870`.
- **[nit]** Consider a `Literal` type for `BuildDiagnostic.level` at `src/apm_cli/marketplace/diagnostics.py:12`
  Only valid values are "warning"/"verbose"; a `Literal` annotation would catch typos at type-check time. Pre-existing gap. DEFERRED (out of scope).

#### CLI Logging Expert

- **[nit]** `compose_codex_marketplace_json` drops the new diagnostic on compose-only callers at `src/apm_cli/marketplace/builder.py:1140`
  The compose-only method returns `(document, warnings)` without diagnostics. The write path (`write_codex_marketplace_json` -> `write_output` -> `compose_output`) surfaces them correctly. DEFERRED (out of scope).

#### DevX UX Expert

- **[nit]** Warning message hardcoded `marketplace.yml` but most users configure in `apm.yml` at `src/apm_cli/marketplace/output_mappers.py:63`
  Retargeted the recovery hint to "your marketplace config". FOLDED in `3a3ba870`.

#### Supply Chain Security Expert

- **[nit]** Many-to-one collision not documented as a known limitation at `src/apm_cli/marketplace/output_mappers.py:20`
  Distinct raw names can map to the same kebab name; not exploitable (output-only, not identity), but worth documenting. FOLDED in `3a3ba870`.

#### OSS Growth Hacker

- **[nit]** CHANGELOG entry buried the user payoff behind implementation detail at `CHANGELOG.md:12`
  Re-led with user value. FOLDED in `3a3ba870`.
- **[nit]** Publish guide did not call out the warning diagnostic at `docs/src/content/docs/producer/publish-to-a-marketplace.md:121`
  Added a sentence naming the rewrite warning. FOLDED in `3a3ba870`.

#### Auth Expert -- inactive

PR touches only marketplace/output_mappers.py, two test files, CHANGELOG.md, and two docs pages -- no auth, token, credential, or remote-host surface affected.

#### Doc Writer

- **[nit]** Sanitisation note lived only on the 7.2 override `name` row; top-level 3.1 `name` was silent at `docs/src/content/docs/reference/manifest-schema.md:809`
  Added a one-clause pointer on the 3.1 `name` row. FOLDED in `3a3ba870`.
- **[nit]** Producer paragraph packed the whole rule into one multi-clause sentence at `docs/src/content/docs/producer/publish-to-a-marketplace.md:118`
  Split the sentence. FOLDED in `3a3ba870`.

#### Supply / Test Coverage Expert

No findings. All behavioral surfaces (Claude name sanitisation, Codex name + displayName preservation, diagnostic emission, no-diagnostic-on-kebab, all-special fallback) have regression-trap tests at integration-with-fixtures tier via the full `MarketplaceBuilder` pipeline. 22 tests pass in 2.14s. Mutation-break gate confirmed.
  *Proof (passed):* `tests/integration/marketplace/test_name_sanitisation_e2e.py::TestClaudeOutputSanitisesName::test_written_json_name_is_kebab_case` -- proves: compiled marketplace.json carries a kebab-case name the Copilot App accepts [devx,portability-by-manifest]
  `assert data["name"] == EXPECTED_NAME`

#### Performance Expert -- inactive

PR touches only marketplace/output_mappers.py, tests, CHANGELOG.md, and docs -- no install, deps, cache, pipeline, resolve, or perf-instrumentation surface.

</details>

<sub>This panel is advisory. It does not block merge. Re-apply the `panel-review` label after addressing feedback to re-run.</sub>
